Framework/Crystal/src/SCDCalibratePanels.cpp

```diff
@@ -605,677 +152,374 @@ void SCDCalibratePanels::exec() {
-      constrain(iFunc, "f0_detHeightScale", MIN_DET_HW_SCALE * detHeightScale0,
-                MAX_DET_HW_SCALE * detHeightScale0);
-      constrain(iFunc, "f0_Xoffset", -1. * maxXYOffset + Xoffset0,
-                maxXYOffset + Xoffset0);
-      constrain(iFunc, "f0_Yoffset", -1. * maxXYOffset + Yoffset0,
-                maxXYOffset + Yoffset0);
-      constrain(iFunc, "f0_Zoffset", -1. * maxXYOffset + Zoffset0,
-                maxXYOffset + Zoffset0);
-
-      double MaxRotOffset = getProperty("MaxRotationChangeDegrees");
-      constrain(iFunc, "f0_Xrot", -1. * MaxRotOffset, MaxRotOffset);
-      constrain(iFunc, "f0_Yrot", -1. * MaxRotOffset, MaxRotOffset);
-      constrain(iFunc, "f0_Zrot", -1. * MaxRotOffset, MaxRotOffset);
-      //} // for vector< string > in Groups
-
-      // Function supports setting the sample position even when it isn't be
-      // refined
-      iFunc->setAttributeValue("SampleX", samplePos.X() + SampleXoffset);
-      iFunc->setAttributeValue("SampleY", samplePos.Y() + SampleYoffset);
-      iFunc->setAttributeValue("SampleZ", samplePos.Z() + SampleZoffset);
-
-      // Constraints for sample offsets
-      if (getProperty("AllowSampleShift")) {
-        maxXYOffset = getProperty("MaxSamplePositionChangeMeters");
-        constrain(iFunc, "SampleX", samplePos.X() + SampleXoffset - maxXYOffset,
-                  samplePos.X() + SampleXoffset + maxXYOffset);
-        constrain(iFunc, "SampleY", samplePos.Y() + SampleYoffset - maxXYOffset,
-                  samplePos.Y() + SampleYoffset + maxXYOffset);
-        constrain(iFunc, "SampleZ", samplePos.Z() + SampleZoffset - maxXYOffset,
-                  samplePos.Z() + SampleZoffset + maxXYOffset);
-      } else {
-        tie(iFunc, true, "SampleX", samplePos.X() + SampleXoffset);
-        tie(iFunc, true, "SampleY", samplePos.Y() + SampleYoffset);
-        tie(iFunc, true, "SampleZ", samplePos.Z() + SampleZoffset);
+    MatrixWorkspace_sptr q3DWS = boost::dynamic_pointer_cast<MatrixWorkspace>(
+        API::WorkspaceFactory::Instance().create(
+            "Workspace2D", 1, 3 * nBankPeaks, 3 * nBankPeaks));
+
+    auto &outSpec = q3DWS->getSpectrum(0);
+    MantidVec &yVec = outSpec.dataY();
+    MantidVec &eVec = outSpec.dataE();
+    MantidVec &xVec = outSpec.dataX();
+
+    for (int i = 0; i < nBankPeaks; i++) {
+      Peak peak = local->getPeak(i);
+      // 1/sigma is considered the weight for the fit
+      double weight = 1.;                // default is even weighting
+      if (peak.getSigmaIntensity() > 0.) // prefer weight by sigmaI
+        weight = 1.0 / peak.getSigmaIntensity();
+      else if (peak.getIntensity() > 0.) // next favorite weight by I
+        weight = 1.0 / peak.getIntensity();
+      else if (peak.getBinCount() > 0.) // then by counts in peak centre
+        weight = 1.0 / peak.getBinCount();
+      for (int j = 0; j < 3; j++) {
+        xVec[i * 3 + j] = i * 3 + j;
+        yVec[i * 3 + j] = 0.0;
```

Comment:
If all values of `yVec` equal 0.0, should we instead set all values at once using `std::fill`?
Similarly, perhaps calculate `i*3+j` once and use that value?